cmd/proofgen/main.go

```diff
@@ -0,0 +1,511 @@
+		desc := fmt.Sprintf("modified proof[%d] bit 3", i)
+		ret = append(ret, inclusionProbe{leafIndex, treeSize, root, leafHash, wrongProof, desc, true})
+	}
+
+	if ln > 0 {
+		ret = append(ret, inclusionProbe{leafIndex, treeSize, root, leafHash, proof[:ln-1], "removed component", true})
+	}
+	if ln > 1 {
+		wrongProof := prepend(proof[1:], proof[0], sha256SomeHash)
+		ret = append(ret, inclusionProbe{leafIndex, treeSize, root, leafHash, wrongProof, "inserted component", true})
+	}
+
+	return ret
+}
+
+func writeCorruptedInclusionTestData(directory string, leafIndex, treeSize uint64, proof [][]byte, root, leafHash []byte) error {
+	happyPath := inclusionProbe{leafIndex, treeSize, root, leafHash, proof, "happy path", false}
+	err := writeInclusionProbe(directory, happyPath)
+	if err != nil {
+		return nil
+	}
+
+	probes := corruptInclusionProof(leafIndex, treeSize, proof, root, leafHash)
+	for _, p := range probes {
+		err = writeInclusionProbe(directory, p)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func writeSingleEntryInclusionTestData(directory string) error {
+	data := []byte("data")
+	// Root and leaf hash for 1-entry tree are the same.
+	hash := hasher.HashLeaf(data)
+	// The corresponding inclusion proof is empty.
+	proof := [][]byte{}
+	emptyHash := []byte{}
+
+	for _, p := range []struct {
+		root    []byte
+		leaf    []byte
+		desc    string
+		wantErr bool
+	}{
+		{hash, hash, "matching root and leaf", false},
+		{hash, emptyHash, "empty leaf", true},
+		{emptyHash, hash, "empty root", true},
+		{emptyHash, emptyHash, "empty root and leaf", true}, // Wrong hash size.
+	} {
+		probe := inclusionProbe{0, 1, p.root, p.leaf, proof, p.desc, p.wantErr}
+
+		err := writeInclusionProbe(directory, probe)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func writeStaticInclusionTestData(rootDirectory string) error {
+	proof := [][]byte{}
+
+	probes := []struct {
+		index, size uint64
+	}{{0, 0}, {0, 1}, {1, 0}, {2, 1}}
+	for i, p := range probes {
+		directory := filepath.Join(rootDirectory, strconv.Itoa(i))
+		err := createDirectory(directory)
+		if err != nil {
+			return err
+		}
+
+		randomLeaf := inclusionProbe{p.index, p.size, []byte{}, sha256SomeHash, proof, "random leaf", true}
+		err = writeInclusionProbe(directory, randomLeaf)
+		if err != nil {
+			return err
+		}
+
+		emptyRoot := inclusionProbe{p.index, p.size, sha256EmptyTreeHash, []byte{}, proof, "empty root", true}
+		err = writeInclusionProbe(directory, emptyRoot)
+		if err != nil {
+			return err
+		}
+
+		emptyRootRandomLeaf := inclusionProbe{p.index, p.size, sha256EmptyTreeHash, sha256SomeHash, proof, "empty root and random leaf", true}
+		err = writeInclusionProbe(directory, emptyRootRandomLeaf)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func writeInclusionProbe(directory string, probe inclusionProbe) error {
+	fileName := strings.Replace(probe.Desc, " ", "-", -1) + ".json"
+
+	probeJson, err := json.Marshal(probe)
```

Comment:
Can we use `json.MarshalIndent` throughout to make the golden json files a bit more (human) readable?

Reply:
v